### DISCUSSION TITLE
[SOLVED] Reified optionals (i.e. optionals assignable to a variable) - a horrifying story

### DISCUSSION THREAD

**Discussion by dumblob** [vlang/v#15015] [General]
**What did you do?**

`x := ?Foo(none)`

**What did you expect to see?**

Compile-time error, that **reified optionals** are not and will never be supported in *V* due to them being nearly useless while opening a can of worms. This has been thoroughly discussed on several places earlier - lately in https://github.com/vlang/v/pull/9567 .
 
**What did you see instead?**

The code compiles.

----

Feel free to describe your experience with reified optionals (like in Rust, Zig, etc.) below.

**Comment by dumblob** [comment]
@medvednikov any update on investigations regarding this important topic happening in the background?

**Comment by medvednikov** [comment]
@dumblob yes, I did some research.

Would you say that by this logic optional fields should also not be allowed?

```
struct Foo {
  s ?string
}
```

And all `?` values are always handled at the call site immediately?

**Comment by medvednikov** [comment]
This would not allow optional functions for example.

See this discussion for context:

https://github.com/vlang/v/discussions/7610#discussioncomment-257982

**Comment by dumblob** [comment]
> @dumblob yes, I did some research.
> 
> Would you say that by this logic optional fields should also not be allowed?

Yes, they should not be allowed.

> ```
> struct Foo {
>   s ?string
> }
> ```
> 
> And all `?` values are always handled at the call site immediately?

Yes.

> This would not allow optional functions for example.
> 
> See this discussion for context:
> 
> [#7610 (reply in thread)](https://github.com/vlang/v/discussions/7610#discussioncomment-257982)

I read the discussion and I think it exhibits kind of a wrong type of thinking. From my understanding the discussion is about runtime safety in case the instance of a callback is not known in compile time and thus can't be assigned in compile time to the struct field. I see two different use cases there:

1. make an instance of the struct and ensure, that the given field is not being read from until it's written to **exactly once**
2. make an instance of the struct and ensure, that the given field is not being read from until it's written to **at least once**

The actual problem though is the fact, that we're trying to accomplish something like this in runtime. There are basically only 2 existing practical solutions to this problem - either the language has something like constructors or the language exposes this problem to the user through a type system. *V* doesn't have any constructors, but coincidentally can express this with its type system - namely with sumtypes (`callback | none`).

And that's also my proposal because it makes it **explicit** and makes it safe to use. This is nothing new as this is how all recursive structures (e.g. trees) in *V* can/shall be defined.

Note, that using sumtypes (`callback | none`) can be also substituted in certain cases by an optimization like the proposed `[required]` in which case *V* language will support one defined pattern checkable in compile time to ensure the use case (2) (because ensuring the use case (1) would require full control flow analysis incl. liveliness which is not what *V* wants_to/can do).

@crthpl @ntrel @elimisteve @danieldaeschle @cristian-ilies-vasile

**Comment by danieldaeschle** [comment]
Using `xyz | none` looks fine to me. And then we need to make a `is` check. But the problem that I see is that we don't support anonymous sum types which would result in a lot of `type` definitions with names like `XyzOrNone`.

**Comment by dumblob** [comment]
>Using `xyz | none` looks fine to me. And then we need to make a `is` check.

Exactly.

>But the problem that I see is that we don't support anonymous sum types which would result in a lot of `type` definitions with names like `XyzOrNone`.

There is always a tradeoff "triangle" which the developer of the given API needs to decide:
```
safety ---------- cognitive load
   \                    /
     \                /
       \            /
       AST verbosity
```

Btw. speaking of anonymous sum types - is there any place where this was discussed?

**Comment by medvednikov** [comment]
@dumblob this is not about callbacks not being known at compile time. It's about having optional callbacks. For example a widget can have 10 callbacks, but only one (say, onclick) defined and used.

Right now in V UI we are doing `if cb != voidptr(0) { cb() }`

**Comment by elimisteve** [comment]
Given the way V currently works, I am confident that the conclusion reached at https://github.com/vlang/v/discussions/7610#discussioncomment-643342 is the best way to go. Requiring that function fields either be (1) optionals or (2) specified when the object is defined is a solution that gives us flexibility, safety, performance, concision ("conciseness"), and with low cognitive load -- a solution that violates the tradeoff triangle that @dumblob has claimed the existence of. (I don't buy this triangle at all; V itself shows us that many so-called "tradeoffs" in language design are made simply because its designers weren't creative enough to come up with a way to have everything we want without significant downsides.)

If we broaden the question to: how _should_ V work, specifically optionals and variants thereof? Then that opens the door for some options :wink: like having not just `?T`, which basically means `T | none | error`, and sometimes splitting this into `T | none` as I think @dumblob may be proposing here, and `T | error`, which is often what we want returned from functions; combining all 3 concepts into Optionals makes some things simpler and other things more complex.

**Comment by elimisteve** [comment]
@dumblob What is your specific solution to having a safe generic tree implementation? Something like this?

```v
struct Node<T> {
    value T
    left Node<T> | none
    right Node<T> | none
}
```

**Comment by danieldaeschle** [comment]
Also correct... Option (short syntax is `?`) should be equal (short) to `Xyz | none`.

**Comment by medvednikov** [comment]
Many people think that Option and Result should be separate, we will most likely separate them.

**Comment by elimisteve** [comment]
@medvednikov I'd be down to allow anonymous sumtypes in certain contexts, such as struct field definitions and function/method parameters, but not defined adhoc inside of functions as new types. This could actually be a significant part of the solution to the question of how exactly -- syntactically and semantically -- V should talk about having Options and Results.

**Comment by elimisteve** [comment]
The limitation could even be so strict as to only allow anonymous sumtypes where one of the "types" is `none`, as in `string | none`. I get that `?T` could eventually mean `T | none`, but `T | none` isn't that verbose and it's super explicit. This is one way to solve the confusion that would stem from the meaning of `?T` changing from `T | none | error` to `T | none`.

**Comment by elimisteve** [comment]
@emily33901 Thoughts on all this? In Discord a couple months ago there were specific situations in which you wanted to be able to pass around optionals that seemed pretty compelling, and that partially convinced me that being able to pass around optionals would be worth the complexity that they'd introduce.

I think we'll remain in good shape as long as errors must be handled immediately and we have the flexibility that https://github.com/vlang/v/discussions/7610#discussioncomment-643342 allows, namely requiring that struct fields of type function must either be set on struct initialization (for safety and performance) or must be `fn | none` (for safety and flexibility) -- no `null` pointers possible! :tada:

**Comment by dumblob** [comment]
> @dumblob What is your specific solution to having a safe generic tree implementation? Something like this?
> 
> ```v
> struct Node<T> {
>     value T
>     left Node<T> | none
>     right Node<T> | none
> }
> ```

Yep.

> Many people think that Option and Result should be separate, we will most likely separate them.

How would Option and Result cope together? Wouldn't this further unnecessarily broaden the (cognitive & syntactic) complexity? Any examples?

> @emily33901 Thoughts on all this? In Discord a couple months ago there were specific situations in which you wanted to be able to pass around optionals that seemed pretty compelling, and that partially convinced me that being able to pass around optionals would be worth the complexity that they'd introduce.

I'm also interested in such examples (I could not come up with any which wouldn't be solved more or less with the same performance and about the same code terseness).

> I think we'll remain in good shape as long as errors must be handled immediately and we have the flexibility that [#7610 (reply in thread)](https://github.com/vlang/v/discussions/7610#discussioncomment-643342) allows, namely requiring that struct fields of type function must either be set on struct initialization (for safety and performance) or must be `fn | none` (for safety and flexibility) -- no `null` pointers possible!

:+1:

**Comment by elimisteve** [comment]
@dumblob Just to make sure: in your ideal world, V would not allow values of type `string | none` or `User | none`, for example, to be passed around, correct?

**Comment by dumblob** [comment]
> @dumblob Just to make sure: in your ideal world, V would not allow values of type `string | none` or `User | none`, for example, to be passed around, correct?

This feels a bit convoluted to me. Let me rephrase it to a question I know an answer to :wink:.

>in your ideal world, V would not allow variables of type `string | none` or `User | none` or `?string`, for example, to be passed around, correct?

Correct (though impossible). But this GitHub issue had a narrower motivation.

Namely I'm fine with `string | none`/`User | none` being assigned to a variable, but I'm not fine with making optionals as well passable around **in addition** to `string | none`. In other words, at the presence of a type-system level mechanism for `... | none`, there must not be any other additional way to express more or less the same thing (the *one way* principle).

My ideal *V* world prefers [row polymorphism](https://journal.infinitenegativeutility.com/structural-types-and-duck-typing ) over subtyping (Crystal lang does that very well - for *V* a simpler variant thereof should be enough though) but with shortcuts e.g. for in-compile-time provable substitution of `string | none` with `string` for performance in cases where possible (e.g. the `[required]` thingy or some smarter static analysis of common cases or whatever).

**Comment by elimisteve** [comment]
There are all kinds of cases where we have nullable values in a database, for example, and having those map to `T | none` (or, at least for now, `?T`), which is exactly the right approach.

If I'm making a parody Facebook clone (or other social media app) and a user hasn't yet specified their profile, then that user's `profile_id` should be `null`. I guess one could generate an empty profile for them, though that is not only wasteful but makes it harder to detect whether the user has filled out their profile or not, and therefore whether I should prompt them to.

In this sort of situation, we can have safety and, uh, have it work by allowing reified optionals:

```v
struct User {
    id int
    username string
    password string
    email string
    profile_id ?int  // maps to nullable int in DB
}

struct UserProfile {
    blood_type string
    sexual_preference SexualPreference
}

enum SexualPreference {  // maps to enum in DB... at least in Postres
    // ...use your imagination
}

fn gen_profile(db DB, profile_id ?int) string { // Key question: how should this work if a `?int` can't be passed in?
    match profile_id {
        none {
            return ''
        }
        int {
            profile := db.select from UserProfile where profile_id = profile_id
            return 'Welcome, ${profile.username}!'
        }
    }
}

...

user := get_user(db) or { // return 404 }
profile_html := gen_profile(user.profile_id) or { "" }
return $vweb.html()
```

I get that vweb typically handles HTML in templates and doesn't have a function generating it, but I hope the point is clear: how can functions like `fn gen_profile(db DB, profile_id ?int) string` exist if we can't pass around `?int` ? And more broadly, how to safely map nullable values from the DB into V?

Reifying values like `int | none` seems like a good idea so as to simply make V practical in such scenarios.

**Comment by dumblob** [comment]
@elimisteve I think we're on the same boat. But I'm just emphasizing, that reifying `?string` is bad unlike reifying `none | string`. These two are completely distinct in *V* (despite on the semantic level one could say, that `?string` is something like `error | string`). And I'm actually fine with reimplementing optionals as syntactic sugar for `error |` sumtype but with the exception, that such instance **must not** be reifiable.

To sum up:

1. `x := (none | string)("abc")` is fine
2. `x := (none | string | error)("abc")` is compile-time error

Another (and maybe simpler to implement and maybe even more performant - I don't know) possibility would be to not merge optionals into sum types and treat them separately (like now) and just disallow reifying these optionals everywhere (making them usable only for returning from functions).

**Comment by elimisteve** [comment]
@dumblob Sounds like you would agree with my database example; great!

Fundamentally, is your point above just that you don't think it's a good idea for V to let people pass around values that must be treated as though they might be errors when they actually can't be errors (as with `x := ?Foo(none)` and many other situations where `T | none` would suffice)?

If so then I agree, and like Alex said, your beef is with Option and Result being combined into `?T`, not with anything specific to function fields on structs, nor compile-time versus runtime setting of values.

FWIW I do like

```v
struct Node<T> {
    value T
    left Node<T> | none
    right Node<T> | none
}
```

though having some syntactic sugar would be a bit nice, like `?T` meaning `T | none` and `!T` meaning `T | error`, which I think someone else proposed in Discord.

`?T` asks, "is T really there or not?", and `!T` says "watch out, this might be an error (if it's not a T)!"

But I still haven't seen a cleaner way to handle such values than `or { ... }` blocks; tricky decisions.

**Comment by dumblob** [comment]
> Fundamentally, is your point above just that you don't think it's a good idea for V to let people pass around values that must be treated as though they might be errors when they actually can't be errors (as with `x := ?Foo(none)` and many other situations where `T | none` would suffice)?

Framing it like this is still not exactly how I envision it. Simply said `T | none` is something wildly different in my eyes than `?T`. `T | none` says something about the semantic value itself and is described by the type (which happens to be a sumtype composed of `none` and `T`). But `?T` says nothing about the semantic value. It says, that control flow took an unexpected/wrong/unwanted/weird/undecidable/another/... path. It doesn't (and mustn't) say anything else. Thus it doesn't make sense to "store" such information in a variable.

So it's not about "in many situations `T | none` would suffice". It's about clear separation of mutually orthogonal concerns. And this separation must be reflected by the syntax.

> If so then I agree, and like Alex said, your beef is with Option and Result being combined into `?T`, not with anything specific to function fields on structs, nor compile-time versus runtime setting of values.

Again, this is somewhat convoluted. Option was always in *V* only about "function returning due to an unexpected/unwanted/... control flow path" until the PR https://github.com/vlang/v/pull/9567 . That suddenly shuffled with the cards a lot and promoted optionals to a very special and very limited sumtype-like mechanism without the need to convolve these two mechanisms.

In that vein, I don't think we need three things (sumtypes, Option, Result) but just two (sumtypes, Option) as it was up until the PR got merged.

> FWIW I do like
> 
> ```v
> struct Node<T> {
>     value T
>     left Node<T> | none
>     right Node<T> | none
> }
> ```
> 
> though having some syntactic sugar would be a bit nice, like `?T` meaning `T | none` and `!T` meaning `T | error`, which I think someone else proposed in Discord.

Yeah, syntactic sugar could be beneficial - though I'd argue not for declaring it but rather for "stripping it off" when used (`match myvar is none {}` might feel verbose; compare it to optionals which have the neat `or {}`). But it must be totally distinct from optionals (i.e. it must not be `?` nor any combination involving `?` at any place).

**Comment by dumblob** [comment]
Hm, thinking more about the syntactic sugar for `none |   ` I think I'd leave it for now without any sugaring. We're already used to sum types being slightly more verbose, but it's manageable and seing `none |   ` among that will IMHO not worsen the situation enough to account for sugar.

And I think so judging based on other languages which do **not have** any nice mechanism for "function returning due to an unexpected/unwanted/... control flow path" and thus are returning `none` instead. Such interfaces are very frequent (most of the world is "impure"). IMHO these languages are the "bad" ones because they convolve two distinct things. *V* did it "right" up until the PR got merged :wink:.

**Comment by elimisteve** [comment]
> I don't think we need three things (sumtypes, Option, Result) but just two (sumtypes, Option)

Can't we make due with just sumtypes + syntactic sugar on top? :cupcake:

**Comment by dumblob** [comment]
> Can't we make due with just sumtypes + syntactic sugar on top?

See my comment above https://github.com/vlang/v/issues/9598#issuecomment-824716508 :wink:.

**Comment by dumblob** [comment]
It seems people try to make up their own `none` in https://github.com/vlang/v/issues/10605#issuecomment-870674595 :open_mouth:.

Maybe the doc should emphasize the use of `| none` instead of inventing `struct None {}; type t0 int | None`.

**Comment by danieldaeschle** [comment]
oh no! :D

**Comment by dumblob** [comment]
This should be solved by RFC https://github.com/vlang/rfcs/pull/7 from @danieldaeschle (assuming my comments in the PR will be included in the RFC :wink:).

**Reply by dumblob** [reply]
https://github.com/vlang/rfcs/pull/7 got implemented and I thus consider this thread as solved (implementation of the RFC revealed the [`or { }` syntax having a dual meaning causes numerous issues](https://github.com/vlang/v/pull/14140#issuecomment-1115944547 ) - but that is not to be discussed here).

Thanks everyone involved!
